File: 2022/06/20_3_sum/yongsu.py
from typing import List
import unittest

# Runtime: 9924 ms, faster than 5.00% of Python3 online submissions for 3Sum.
# Memory Usage: 18 MB, less than 71.84% of Python3 online submissions for 3Sum.

"""
배열 C로 관리하기 어려워서 python으로 함.. C 나중에 연습하자
"""

"""
1. 오름차순으로 정렬
2. 숫자를 고정
3. 두 개의 인덱스 (left right)를 통해 합이 0이 되는 조합을 찾는다 (left가 right와 같을때까지) 이때 left i를 제외한 가장 왼쪽부터, right는 가장 오른쪽부터 시작
4. 이 때, 배열이 오름차순이기 때문에 세 수의 합이 0보다 작으면 무조건 left를 한 칸 이동한다. 반대로 세 수의 합이 0보다 크면 right를 한 칸 땡긴다
5. 중복 제거를 위해 다음 숫자가 같으면 한 칸씩 더 땡긴다 (아이디어만 생각했는데 버그를 못잡아서 이미 있는 조합이면 안 넣는 것으로 변경)
"""
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()

        for i in range(len(nums) - 2):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                if sum < 0:
                    left += 1
                elif sum > 0:
                    right -= 1
                else:
                    tmp = [nums[i], nums[left], nums[right]]
                    if tmp not in ans:
                        ans.append(tmp)
                    left += 1
                    right -= 1

        return ans


class SolutionTests(unittest.TestCase):
    def setUp(self) -> None:
        self.sol = Solution()

    def test_1(self):
        nums = [-1, 0, 1, 2, -1, -4]
        answer = [[-1, -1, 2], [-1, 0, 1]]
        result = self.sol.threeSum(nums)
        self.assertEqual(result, answer)


if __name__ == "__main__":
    unittest.main()

File: 2022/06/28_search_in_rotated_sorted_array/yongsu.c
#include <stdio.h>


// Runtime: 6 ms, faster than 44.46% of C online submissions for Search in Rotated Sorted Array.
// Memory Usage: 6.2 MB, less than 7.51% of C online submissions for Search in Rotated Sorted Array.

// 이진탐색인데 pivot 기준으로 값 비교하면서 탐색하기
// pivot 찾는 것도 이진탐색으로
int search(int* nums, int numsSize, int target){
    int i;
    int pivot = 0;
    int original_array[numsSize + 10];
    int left = 0;
    int right = numsSize - 1;
    int mid;

    while (left < right) {
        mid = (left + right) / 2;
        if (nums[mid] > nums[right]) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    pivot = left;

    left = 0;
    right = numsSize - 1;

    int real_mid;

    while (left <= right) {
        mid = (left + right) / 2;
        real_mid = mid + pivot > numsSize - 1 ? mid + pivot - numsSize : mid + pivot; 
        if (target == nums[real_mid]) {
            return real_mid;
        } else if (target < nums[real_mid]) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }   

    return -1; 
}

int main(void) {
    printf("%d\n", search((int []){4, 5, 6, 7, 0, 1, 2}, 7, 0)); // 4
    return 0;
}